ethcore/machine/src/executive.rs

```diff
@@ -435,9 +451,16 @@ impl<'a> CallCreateExecutive<'a> {
 							})
 						}
 					} else {
+						// EIP161 - reverted calls cancels state trie cleaning
+						if initial_builtin_balance.is_zero() {
+							let prune = UNPRUNABLE_PRECOMPILE_ADDRESS
+								.map_or(true, |addr| addr != params.code_address);
+							if prune {
+								substate.touched.remove(&params.code_address);
+							} 
+						}
```

Comment:
just to make sure I understand this correctly: for tests this will remove all but ripemd builtints with zero balance and for production this revert all zero-balanced builtins (although in practice all builtins have `balance: 1`, maybe there are some exceptions https://github.com/openethereum/openethereum/pull/11482/files#r378702588, this needs to be double checked)

Reply:
That's a great observation, `classic` and `foundation` have `balance == 0` more might have it too.

Reply:
> for tests this will remove all but ripemd builtints with zero balance

we need to be able to reproduce the RIPEMD removal at block `#2686351` in tests

> for production this revert all zero-balanced builtins 

afaik, all precompiles has `balance: 1` due the EIP161 rule. for instance, when you create a new clique genesis with geth, all precompiles balances are set to `1`

> maybe there are some exceptions

this issue happened due that both geth and parity allowed the removal of 0x3 account for different reasons, and nobody noticed about it after some time, so core devs decided to introduce this "irregular state transition". 

I agree with you that this change could affect to PoA chains with the same scenario:
- they initially specified the precompile in the trie with `balance=0`
- they called a precompile without enough gas and the precompile reverted and was removed from the trie

Maybe we can add there a `eip161_touch_prunning_bug`, but this is really ugly :scream:

Reply:
Moving everything to chain definitions is another option,

- `eip161_unprunable_precompiles`: list of addresses that cannot be pruned if touched with zero balance.

we have to add this flag to the testnet with `0x3`, and also allows fixing existing PoA chains. I think that we need to generate some kind of an explicit `warn!` log there to help to identity the problem.

Reply:
@ordian @niklasad1,

I see these options now:

- (a) Fix EIP161 only in tests: in real deployments, this is "tactically fixed" by setting balances to >=1 in precompile accounts.
- (b) Fix EIP161 at all
  - (b.1) add a chain option `eip161_unprunable_precompiles=addr, addr,...` and define it in mainnet chains with 0x3
  - (b.2) add a chain option `eip161_touch_prunning_bug`, a flag that could be activated by any existing PoA chains that removed some precompile accounts
- (c) Introduce generic irregular state transitions (I do not like this)

Any other idea? Thoughts?

Reply:
I would go for option (a) as the easiest one, but we would need to ensure (somehow via CI and document it enough) that each precompile has balance >= 1 indeed. The problem with changing balance for existing chains is that it's a breaking change (see e.g. #11108).

Reply:
@ordian, implemented (a) in this PR https://github.com/openethereum/openethereum/commit/04dc2673651dd0019de31a457ab153d4ba03cd93